subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -405,9 +755,222 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	}
+
+	bond = fp_bond_info_free_get();
+	if (!bond) {
+		return -ENOMEM;
+	}
+
+	bond->conn_ctx = conn_ctx;
+	bond->bonded = false;
+	bond->nv.account_key_metadata_id = id;
+	bond->nv.addr = *addr;
+
+	err = fp_bond_info_settings_save(bond, &bond_rollback_copy);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+void fp_storage_ak_bond_conn_confirm(const void *conn_ctx, const bt_addr_le_t *addr)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	if (!is_enabled) {
+		return;
+	}
+
+	struct fp_bond_info *bond_confirmed;
+	int err;
+
+	__ASSERT_NO_MSG(bt_addr_le_is_identity(addr));
+
+	bond_confirmed = fp_bond_info_by_conn_get(conn_ctx);
+	if (!bond_confirmed) {
+		LOG_ERR("Bond not found in Fast Pair storage");
+		return;
+	}
+
+	bond_confirmed->bonded = true;
+
+	/* Remove bonded duplicates that were overwritten by new bond. */
+	FP_BONDS_FOREACH(bond) {
+		if ((bond->conn_ctx != conn_ctx) && bond->bonded &&
+		    bt_addr_le_eq(&bond->nv.addr, addr)) {
+			/* Only one bonded duplicate can be present here. */
+			err = bond_remove(bond, false);
+			if (err) {
+				LOG_ERR("bond_remove failed (err %d)", err);
+				break;
+			}
+
+			LOG_DBG("Removed duplicated bond");
+			break;
+		}
+	}
+
+	if (bond_confirmed->nv.account_key_metadata_id != FP_BOND_INFO_AK_ID_UNKNOWN) {
+		/* Account Key already associated (subsequent pairing procedure).
+		 * Procedure finished successfully. Setting conn_ctx to NULL.
+		 */
+		bond_confirmed->conn_ctx = NULL;
+	}
+}
+
+void fp_storage_ak_bond_conn_addr_update(const void *conn_ctx, const bt_addr_le_t *new_addr)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	if (!is_enabled) {
+		return;
+	}
+
+	int err;
+	struct fp_bond_info *bond;
+	struct fp_bond_info bond_rollback_copy;
+
+	bond = fp_bond_info_by_conn_get(conn_ctx);
+	if (!bond) {
+		return;
+	}
+	bond_rollback_copy = *bond;
+
+	bond->nv.addr = *new_addr;
+
+	err = fp_bond_info_settings_save(bond, &bond_rollback_copy);
+	if (err) {
+		return;
+	}
+}
+
+void fp_storage_ak_bond_conn_finalize(const void *conn_ctx)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	if (!is_enabled) {
+		return;
+	}
+
+	struct fp_bond_info *bond;
+	int err;
+
+	bond = fp_bond_info_by_conn_get(conn_ctx);
+	if (!bond) {
+		return;
+	}
+
+	/* If at this point the conn_ctx is not set to NULL, the procedure is incomplete. */
+	err = bond_remove(bond, true);
+	if (err) {
+		LOG_ERR("bond_remove failed (err %d)", err);
+	}
+}
+
+void fp_storage_ak_bond_delete(const bt_addr_le_t *addr)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	if (!is_enabled) {
+		return;
+	}
+
+	LOG_DBG("Bluetooth bond deleted");
+	FP_BONDS_FOREACH(unused) {
+		struct fp_bond_info *bond;
+		int err;
+
+		/* In some edge cases, while enabling and validating bonds there might be more than
```

Comment:
Those loops call bt_unpair which calls this function callback so we can land here when handling bond duplicates during Fast Pair enable operation. We can't firstly remove the bond duplicates from the Fast Pair storage because of power down risk.

Reply:
You are right. Let's keep it as is.